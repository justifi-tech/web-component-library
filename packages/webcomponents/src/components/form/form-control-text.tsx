import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  State,
  Element,
  Watch,
} from '@stencil/core';

@Component({
  tag: 'form-control-text'
})
export class TextInput {
  @Element() el: HTMLElement;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() defaultValue: string;
  @Prop() maxLength?: number;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() keyDownHandler?: (event: any) => void;
  @Prop() disabled: boolean;

  @State() input: string;

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  updateInput(newValue: any) {
    const inputElement = this.el.querySelector('input');
    if (inputElement) {
      inputElement.value = newValue || '';
    }
  }

  handleFormControlInput(event: any) {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
    this.formControlInput.emit(target.value);
  }

  componentDidLoad() {
    this.updateInput(this.defaultValue);
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <label part="label" class="form-label" htmlFor={this.name}>
          {this.label}
        </label>
        <input
          id={this.name}
          name={this.name}
          onInput={(event: any) => this.handleFormControlInput(event)}
          onBlur={() => this.formControlBlur.emit()}
          onKeyDown={(event: any) => this.keyDownHandler && this.keyDownHandler(event)}
          maxLength={this.maxLength}
          part={`input ${this.errorText ? 'input-invalid ' : ''}${this.disabled ? ' input-disabled' : ''}`}
          class={this.errorText ? 'form-control is-invalid' : 'form-control'}
          type="text"
          disabled={this.disabled}
        />
        <form-control-help-text helpText={this.helpText} />
        <br />
        <form-control-error-text errorText={this.errorText} />
      </Host>
    );
  }
}

