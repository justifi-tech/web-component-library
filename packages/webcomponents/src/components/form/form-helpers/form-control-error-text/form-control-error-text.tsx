import { Component, Prop, h } from '@stencil/core';

@Component({
  tag: 'form-control-error-text',
})
export class FormControlErrorText {
  @Prop() errorText: string;
  @Prop() name: string;

  render() {

    if (!this.errorText) {
      return null;
    }

    return (
      <small class='form-text text-danger' id={`form-error-text-${this.name}`}>
        {this.errorText}
      </small>
    );
  }
}
