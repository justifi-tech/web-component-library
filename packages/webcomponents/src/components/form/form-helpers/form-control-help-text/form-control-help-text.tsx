import { Component, Prop, h } from '@stencil/core';

@Component({
  tag: 'form-control-help-text',
})
export class FormControlHelpText {
  @Prop() helpText: string;
  @Prop() errorText: string;

  private text: string;

  get colorClass() {
    return this.errorText ? 'text-danger' : 'text-muted';
  }

  render() {

    if (!this.helpText && !this.errorText) {
      return null;
    }

    if (this.errorText) {
      this.text = this.errorText;
    } else {
      this.text = this.helpText;
    }

    return (
      <small class={`form-text ${this.colorClass}`} id='formHelpText'>
        {this.text}
      </small>
    );
  }
}
