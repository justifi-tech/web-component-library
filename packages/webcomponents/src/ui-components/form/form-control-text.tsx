import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Element,
  Watch,
} from '@stencil/core';

@Component({
  tag: 'form-control-text'
})
export class TextInput {
  @Element() el: HTMLElement;

  @Prop() name: any;
  @Prop() label: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() defaultValue?: string;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() maxLength?: number;
  @Prop() keyDownHandler?: (event: any) => void;
  @Prop() disabled?: boolean;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  componentDidLoad() {
    this.updateInput(this.defaultValue);
  }

  handleFormControlInput = (event: any) => {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
    this.formControlInput.emit({ name, value: target.value });
  }

  updateInput = (newValue: any) => {
    const inputElement = this.el.querySelector('input');
    if (inputElement) {
      inputElement.value = newValue || '';
    }
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <div class="form-group d-flex flex-column">
          <div class="d-flex align-items-start gap-2">
            <label part="label" class="form-label" htmlFor={this.name}>
              {this.label}
            </label>
            <form-control-tooltip helpText={this.helpText} />
          </div>
          <input
            id={this.name}
            name={this.name}
            onBlur={this.formControlBlur.emit}
            onInput={this.handleFormControlInput}
            onKeyDown={this.keyDownHandler}
            onPaste={this.keyDownHandler}
            maxLength={this.maxLength}
            part={`input ${this.errorText ? "input-invalid " : ""}${this.disabled ? " input-disabled" : ""}`}
            class={this.errorText ? "form-control is-invalid" : "form-control"}
            type="text"
            disabled={this.disabled}
          />
          <form-control-error-text errorText={this.errorText} name={this.name} />
        </div>
      </Host>
    );
  }
}
