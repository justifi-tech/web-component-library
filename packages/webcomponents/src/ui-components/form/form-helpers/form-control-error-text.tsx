import { h } from '@stencil/core';

export const FormControlErrorText = (props: { errorText: string, name: string }) => {
  const { errorText, name } = props;

  if (!errorText) {
    return null;
  }

  return (
    <small class='form-text text-danger' id={`form-error-text-${name}`} part="input-error-text" >
      {errorText}
    </small>
  );
}
