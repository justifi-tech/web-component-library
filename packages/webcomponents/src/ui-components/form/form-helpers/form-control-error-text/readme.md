# form-control-error-text



<!-- Auto Generated Below -->


## Properties

| Property    | Attribute    | Description | Type     | Default     |
| ----------- | ------------ | ----------- | -------- | ----------- |
| `errorText` | `error-text` |             | `string` | `undefined` |
| `name`      | `name`       |             | `string` | `undefined` |


## Dependencies

### Used by

 - [form-control-checkbox](../..)
 - [form-control-date](../..)
 - [form-control-file](../..)
 - [form-control-file-v2](../..)
 - [form-control-monetary](../..)
 - [form-control-number](../..)
 - [form-control-number-masked](../..)
 - [form-control-select](../..)
 - [form-control-text](../..)
 - [form-control-textarea](../..)
 - [justifi-dispute-reason](../../../../components/dispute-management/dispute-response)

### Graph
```mermaid
graph TD;
  form-control-checkbox --> form-control-error-text
  form-control-date --> form-control-error-text
  form-control-file --> form-control-error-text
  form-control-file-v2 --> form-control-error-text
  form-control-monetary --> form-control-error-text
  form-control-number --> form-control-error-text
  form-control-number-masked --> form-control-error-text
  form-control-select --> form-control-error-text
  form-control-text --> form-control-error-text
  form-control-textarea --> form-control-error-text
  justifi-dispute-reason --> form-control-error-text
  style form-control-error-text fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
