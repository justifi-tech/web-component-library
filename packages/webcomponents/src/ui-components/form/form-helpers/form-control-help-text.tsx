import { h } from '@stencil/core';

export const FormControlHelpText = (props: { helpText: string, name: string }) => {
  const { helpText, name } = props;

  if (!helpText) {
    return null;
  }

  return (
    <small class='form-text text-muted' id={`form-help-text-${name}`} part="input-help-text">
      {helpText}
    </small>
  );
}
