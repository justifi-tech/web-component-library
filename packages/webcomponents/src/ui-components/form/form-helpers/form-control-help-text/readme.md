# form-control-help-text



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute   | Description | Type     | Default     |
| ---------- | ----------- | ----------- | -------- | ----------- |
| `helpText` | `help-text` |             | `string` | `undefined` |
| `name`     | `name`      |             | `string` | `undefined` |


## Dependencies

### Used by

 - [bank-account-form-inputs](../../../../components/business-forms/payment-provisioning/bank-account)
 - [bank-account-form-inputs-canada](../../../../components/business-forms/payment-provisioning/bank-account)
 - [form-control-radio](../..)
 - [form-control-textarea](../..)

### Graph
```mermaid
graph TD;
  bank-account-form-inputs --> form-control-help-text
  bank-account-form-inputs-canada --> form-control-help-text
  form-control-radio --> form-control-help-text
  form-control-textarea --> form-control-help-text
  style form-control-help-text fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
