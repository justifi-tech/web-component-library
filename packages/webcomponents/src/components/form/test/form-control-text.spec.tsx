import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { TextInput } from '../form-control-text';
import { FormControlErrorText } from '../form-helpers/form-control-error-text/form-control-error-text';
import { FormControlHelpText } from '../form-helpers/form-control-help-text/form-control-help-text';

describe('form-control-text', () => {
  const components = [TextInput, FormControlErrorText, FormControlHelpText];
  const mockInputHandler = jest.fn();
  const mockKeyDownHandler = jest.fn();

  it('Renders with default props', async () => {
    const page = await newSpecPage({
      components: components,
      template: () => <form-control-text label='Username' name='username'></form-control-text>,
    });

    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.label).toBe('Username');
    expect(page.rootInstance.name).toBe('username');
  });

  it('Renders with all props provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-text
          label='Email'
          name='email'
          defaultValue='user@example.com'
          errorText='Invalid email'
          helpText='Enter your email'
          disabled
          maxLength={50}
          inputHandler={mockInputHandler}
          keyDownHandler={mockKeyDownHandler}
        >
        </form-control-text>
    });

    expect(page.root).toMatchSnapshot();
  });

  it('Updates input value on defaultValue prop change', async () => {
    // Initial render with the default value
    let defaultValue = 'user@example.com';
    let page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-text
          label='Email'
          name='email'
          defaultValue={defaultValue}
        >
        </form-control-text>
    });

    let inputElement = page.root.querySelector('input');
    expect(inputElement.value).toBe('user@example.com');

    defaultValue = 'updated@example.com';

    // Re-render component with new defaultValue
    page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-text
          label='Email'
          name='email'
          defaultValue={defaultValue}
        >
        </form-control-text>
    });

    inputElement = page.root.querySelector('input');
    expect(inputElement.value).toBe('updated@example.com');
  });

  it('Handles user input correctly', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-text
          label='Email'
          name='email'
          defaultValue='user@example.com'
          inputHandler={mockInputHandler}
        >
        </form-control-text>
    });

    const inputElement = page.root.querySelector('input');
    const testValue = 'Hello, World!';

    inputElement.value = testValue;
    await inputElement.dispatchEvent(new Event('input'));

    expect(inputElement.value).toBe(testValue);
  });

  it('Emits formControlInput event on input', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-text
          label='Email'
          name='email'
          defaultValue='user@example.com'
          inputHandler={mockInputHandler}
        >
        </form-control-text>
    });

    const inputEventSpy = jest.fn();
    page.root.addEventListener('formControlInput', inputEventSpy);

    await page.waitForChanges();

    const inputElement = page.root.querySelector('input');
    inputElement.value = 'Hello, World!';

    inputElement.dispatchEvent(new Event('input', { bubbles: true, composed: true }));

    await page.waitForChanges();

    expect(inputEventSpy).toHaveBeenCalled();
  });

  it('Emits formControlBlur event on blur', async () => {
    const page = await newSpecPage({
      components: components,
      template: () => <form-control-text defaultValue='user@example.com' inputHandler={mockInputHandler}></form-control-text>
    });

    const blurEventSpy = jest.fn();
    page.root.addEventListener('formControlBlur', blurEventSpy);

    const inputElement = page.root.querySelector('input');
    inputElement.dispatchEvent(new Event('blur'));

    expect(blurEventSpy).toHaveBeenCalled();
  });

  it('Disables input when disabled prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () => <form-control-text label='Username' name='username' disabled></form-control-text>,
    });

    const inputElement = page.root.querySelector('input');
    expect(inputElement.disabled).toBeTruthy();
  });

  it('Shows help text when helpText prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () => <form-control-text label='Username' name='username' helpText='This is the help text.'></form-control-text>,
    });

    const helpTextComponent = page.root.querySelector('form-control-help-text');
    expect(helpTextComponent).not.toBeNull();

    const helpText = helpTextComponent.querySelector('.text-muted');
    expect(helpText.textContent).toBe('This is the help text.');
  });

  it('Shows error and applies error styling when error prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () => <form-control-text label='Username' name='username' errorText='This field is required.'></form-control-text>,
    });

    const errorTextComponent = page.root.querySelector('form-control-error-text');
    expect(errorTextComponent).not.toBeNull();

    const errorText = errorTextComponent.querySelector('.text-danger');
    expect(errorText.textContent).toBe('This field is required.');
  });
});
