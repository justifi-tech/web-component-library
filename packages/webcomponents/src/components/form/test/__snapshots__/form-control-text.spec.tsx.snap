// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form-control-text renders with default props 1`] = `
<form-control-text exportparts="label,input,input-invalid" label="Username" name="username">
  <label class="form-label" htmlfor="username" part="label">
    Username
  </label>
  <input class="form-control" id="username" name="username" part="input " type="text" value="">
  <form-control-help-text></form-control-help-text>
  <br>
  <form-control-error-text></form-control-error-text>
</form-control-text>
`;